metaflow/metaflow_config_funcs.py

```diff
@@ -0,0 +1,120 @@
+        )
+    return config
+
+
+# Initialize defaults required to setup environment variables.
+METAFLOW_CONFIG = init_config()
+
+_all_configs = {}
+
+
+def config_values(include=0):
+    # By default, we just return non-null values that are marked as "propagate" and that
+    # are not default. This is the common use case because in all other cases, the code
+    # is sufficient to recreate the value (ie: there is no external source for the value)
+    for name, config_value in _all_configs.items():
+        if (config_value.value is not None or include & NULL_VALUES) and (
+            not config_value.is_default or include & NON_CHANGED_VALUES
+        ):
+            yield name, config_value.serializer(config_value.value)
+
+
+def from_conf(name, default=None, validate_fn=None):
+    """
+    First try to pull value from environment, then from metaflow config JSON
+
+    Prior to a value being returned, we will validate using validate_fn (if provided).
+    Only non-None values are validated.
+
+    validate_fn should accept (name, value).
+    If the value validates, return None, else raise an MetaflowException.
+    """
+    env_name = "METAFLOW_%s" % name
+    is_default = True
+    value = os.environ.get(env_name, METAFLOW_CONFIG.get(env_name, default))
+    if validate_fn and value is not None:
+        validate_fn(env_name, value)
+    if default is not None:
+        # In this case, value is definitely not None because default is the ultimate
+        # fallback and all other cases will return a string (even if an empty string)
+        if isinstance(default, (list, dict)):
+            # If we used the default, value is already a list or dict, else it is a
+            # string so we can just compare types to determine is_default
+            if isinstance(value, (list, dict)):
+                is_default = True
+            else:
+                try:
+                    value = json.loads(value)
+                except json.JSONDecodeError:
+                    raise ValueError(
+                        "Expected a valid JSON for %s, got: %s" % (env_name, value)
+                    )
+            _all_configs[env_name] = ConfigValue(
+                value=value,
+                serializer=json.dumps,
+                is_default=is_default,
+            )
+            return value
+        elif isinstance(default, (bool, int, float)) or is_stringish(default):
+            try:
+                value = type(default)(value)
+                # Here we can compare values
+                is_default = value == default
+            except ValueError:
+                raise ValueError(
+                    "Expected a %s for %s, got: %s" % (type(default), env_name, value)
+                )
+        else:
+            raise RuntimeError(
+                "Default of type %s for %s is not supported" % (type(default), env_name)
+            )
+    _all_configs[env_name] = ConfigValue(
+        value=value,
+        serializer=str,
```

Comment:
are we sure that the serializer has to be `str` always? Currently, in the tip of master - there are a few config values that are cast to int/bool.

Reply:
see below comment but yes, we deserialize from str to the type (int/bool) and serialize back to str when emitting back out as env var. It's either str or json.loads/dumps